keps/sig-api-machinery/20190324-declatative-field-manager.md

```diff
@@ -0,0 +1,207 @@
+
+To keep the way of interacting with manifests as easy as possible, this KEP suggests to allow defining the fieldManager as part of the manifest itself. Therefore no further action is required when applying a set of manifests and users can easily benefit from Server Side Apply without having to change their apply workflow.
+
+This would also assist existing solutions already interacting with manifest files, as the fieldManager information gets provided in a way they already know to handle which might improve compatibility.
+
+### Goals
+
+- Allow users to set their fieldManager information declaratively in their manifests
+
+### Non-Goals
+
+- Replace the existing ways to set the fieldManager information
+
+## Proposal
+
+To achieve the goals, we would add a field to ObjectMeta called `options`, this field would allow to send options to the api server.
+The field will be in the form of a map of string keys and values that gets validated to contain only known options (currently `fieldManager`).
+
+This field should not get persisted to storage and is solely for sending request options information to the apiserver.
+When the `options.fieldManager` value is not set or empty, the apiserver would fallback onto current behavior and default the fieldManager (or fail).
+
+This means, we add a field that is optional metadata, write-only and non-persisted and exists for the sole purpose of sending request information to the apiserver.
+The `options.fieldManager` value should accept a non empty string or be unset and should follow the same criteria as [ManagedFieldsEntry.Manager](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.14/#managedfieldsentry-v1-meta).
+
+Setting or not setting the `options.fieldManager` value will cause the following behavior for both apply and non-apply operations:
+
+- If the fieldManager request option is set for the request (for example through kubectl), it will get used as manager
+- If the newly introduced `options.fieldManager` option is set in the received request, it will get used as manager
+- If both of the two options are set and do not match, the request will get rejected.
+
+An example of setting the field would be:
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: example
+  options:
+    fieldManager: jenkins
+data:
+  k: v
+```
+
+The challenge of this proposal is to introduce a new kind of field behaving differently from the fields currently available in the api,
+as it does not get persisted directly, but causes side-effects on the apiserver.
+
+### Risks and Mitigations
+
+One risk might be, that introducing this change would mean there is a field that does not get persisted to storage.
+As a result, setting it causes different behavior on the apiserver, but is not reflected when reading the object back from the apiserver.
```

Comment:
As far as I can tell, the field would be sent as part of the manifest (the first time), and also saved in the last-applied annotation. The following times, since it would be missing from the server but present both in the config and the last-applied annotation, the field wouldn't be sent? Though I'm not completely sure about this behavior.

Reply:
I'm not sure if that behavior is super critical either though.